lib/puppet/util.rb

```diff
@@ -349,6 +356,52 @@ def uri_to_path(uri)
   end
   module_function :uri_to_path
 
+  private
+
+  RFC_3986_URI_REGEX = /^(?<scheme>([^:\/?#]+):)?(?<authority>\/\/([^\/?#]*))?(?<path>[^?#]*)(\?(?<query>[^#]*))?(#(?<fragment>.*))?$/
+
+  public
+
+  # Will escape a query parameter or any portion of the query string
+  # so that it will remain unescapable with URI.unescape
+  def uri_query_escape(query_string)
+    return nil if query_string.nil?
+    # query can encode space to %20 OR +
+    # + MUST be encoded as %2B
+    # CGI.escape turns space into + which is the most backward compatible
+    # however it doesn't roundtrip through URI.unescape which prefers %20
+    CGI.escape(query_string).gsub('+', '%20')
+  end
+  module_function :uri_query_escape
+
+  # Takes a string containing characters that should be encoded in a Uri
+  # The output is safe to pass to URI.parse or URI::Generic.build  because it
+  # unescapes characters / and : prior to producing a value
+  def uri_escape(path, opts = { :allow_fragment => false })
+    raise ArgumentError.new('path may not be nil') if path.nil?
+
+    escaped = ''
+
+    # parse uri into named matches, then reassemble properly escaped
+    parts = path.match(RFC_3986_URI_REGEX)
+
+    escaped += parts[:scheme] unless parts[:scheme].nil?
+    escaped += parts[:authority] unless parts[:authority].nil?
+
+    # path requires space to be encoded as %20 (NEVER +)
+    # + should be left unencoded
+    # URI::parse and URI::Generic.build don't like values from CGI.escape(path)
+    # URI.escape does not change / to %2F and : to %3A like CGI.escape
+    escaped += URI.escape(parts[:path]) unless parts[:path].nil?
+
+    escaped += ('?' + uri_query_escape(parts[:query])) unless parts[:query].nil?
+    # TODO: does fragment have same rules?
```

Comment:
I believe fragment follows the same rules as query... and commits should be tweaked accordingly